src/Stratis.Bitcoin.Tests/Base/ConsensusManagerBehaviorTests.cs

```diff
@@ -0,0 +1,278 @@
+            chain.SetTip(consensusTip);
+
+            // Ibd
+            var ibdState = new Mock<IInitialBlockDownloadState>();
+            ibdState.Setup(x => x.IsInitialBlockDownload()).Returns(() => this.IsIBD);
+
+            // Consensus manager
+            var cmMock = new Mock<IConsensusManager>();
+
+            cmMock.Setup(x => x.HeadersPresented(It.IsAny<INetworkPeer>(), It.IsAny<List<BlockHeader>>(), It.IsAny<bool>()))
+                .Returns((INetworkPeer p, List<BlockHeader> presentedHeaders, bool triggerDownload) =>
+            {
+                this.headersPresentedCalledTimes++;
+
+                return connectNewHeadersMethod?.Invoke(presentedHeaders, triggerDownload);
+            });
+
+            cmMock.Setup(x => x.Tip).Returns(consensusTip);
+
+            var cmBehavior = new ConsensusManagerBehavior(chain, ibdState.Object, cmMock.Object, new Mock<IPeerBanning>().Object,
+                new Mock<IConnectionManager>().Object, this.loggerFactory);
+
+            // Peer and behavior
+            this.peerMock = this.CreatePeerMock();
+
+            cmBehavior.Attach(this.peerMock.Object);
+            this.peerMock.Setup(x => x.Behavior<ConsensusManagerBehavior>()).Returns(() => cmBehavior);
+            this.peerMock.Setup(x => x.State).Returns(peerState);
+
+            if (expectedPeerTip != null)
+                cmBehavior.SetPrivatePropertyValue("ExpectedPeerTip", expectedPeerTip);
+
+            if (cache != null)
+                cmBehavior.SetPrivateVariableValue("cachedHeaders", cache);
+
+            var counter = new Counter();
+
+            this.peerMock.Setup(x => x.SendMessageAsync(It.IsAny<Payload>(), It.IsAny<CancellationToken>())).Returns((Payload payload, CancellationToken token) =>
+            {
+                if (payload is GetHeadersPayload)
+                    counter.Value++;
+
+                return Task.CompletedTask;
+            });
+
+            getHeadersPayloadSentTimes = counter;
+
+            return cmBehavior;
+        }
+
+        private List<BlockHeader> getCachedHeaders(ConsensusManagerBehavior behavior)
+        {
+            return behavior.GetMemberValue("cachedHeaders") as List<BlockHeader>;
+        }
+
+        /// <summary>
+        /// CT is at 5. peer 1 claims block 10 (<see cref="ConsensusManagerBehavior.ExpectedPeerTip"/> is header10).
+        /// Cached headers contain nothing. <see cref="ConsensusManagerBehavior.ConsensusTipChangedAsync"/> called with header 6.
+        /// Make sure that getHeaders payload wasn't sent to the peer, <see cref="ConsensusManager.HeadersPresented"/> wasn't called. Return value is <c>null</c>.
+        /// </summary>
+        [Fact]
+        public async Task ConsensusTipChanged_CTAdvancedBuNoCachedHeadersAsync()
+        {
+            ConsensusManagerBehavior behavior = this.CreateAndAttachBehavior(this.headers[5], out Counter getHeadersPayloadSentTimes, null, this.headers[10]);
```

Comment:
It is little bit confusing when the comment says block 10 or header 10. Is it the block with height of 10? Because otherwise the block number 10 would be at index nine in the array

Reply:
I checked the creation of the list of headers and you don't have Genesis in it which means all your indexes are off by one

Reply:
Added genesis header as the first item. 

Good point, but this doesnt really affect the logic